### DISCUSSION TITLE
v5: http logger

### DISCUSSION THREAD

**Discussion by adufr** [adonisjs/core#1500] [Help] [answered]
Hi!

I was previously using [adonis-http-logger](https://github.com/adonisjs/adonis-http-logger) on my v4 application, and since the package hasn't been migrated yet, I want to implement it myself on my Adonis v5 API.

However, I don't know what could be the equivalent of the following in Adonis v5
```js
      HttpContext.onReady(function (ctx) {
        const logger = new Logger(ctx, AdonisLogger)
        logger.hook()
      })
```

Can someone help me figure this out? :)

**Comment by popovicieduard** [comment, marked as answer]
I think you can add a .ts file inside /start and preload it inside .adonisrc.json
```
  "preloads": [
    "./start/routes",
    "./start/kernel",
    "./start/events",
    "./start/validationRules",
    "./start/yourfile"
  ],
```

Don't know if you can pass ctx inside preloads files

**Reply by adufr** [reply]
Even though I could access the HttpContext, there are no `onReady()` method where I could hook my logger like there was in Adonis v4

**Reply by skrenek** [reply]
This is working for me in a preload file.

    /* eslint-disable max-len */
    import Server from '@ioc:Adonis/Core/Server'
    import {DateTime} from 'luxon'
    
    const DATE_FORMAT = 'dd/MMM/yyyy:HH:mm:ss ZZZ'
    
    Server.hooks.after(async (ctx) => {
      ctx.logger.info(`${ctx.request.ip()} - ${ctx.request.userId || '-'} [${DateTime.local().toFormat(DATE_FORMAT)}] "${ctx.request.method()} ${ctx.request.url(true)} HTTP/${ctx.request.request.httpVersion || '-'}" ${ctx.response.response.statusCode} - ${ctx.request.header('referrer', '-')} "${ctx.request.header('user-agent', '-')}"`)
    })

**Reply by popovicieduard** [reply]
I think Woosy is trying to catch errors with a global logger, @Woosy why aren't you using the report method from the Exception/Handler.ts ? https://preview.adonisjs.com/guides/http/exception-handling#the-report-method

**Reply by adufr** [reply]
@skrenek your solution worked perfectly! I was looking for this function (`Server.hooks.after`).
The problem I had with middleware was that my logging function wasn't called when there was an error (like with a 400 status code), and I really want to log ALL requests into my console.
Anyway topic is now solved! Thank you all :+1:

**Reply by skrenek** [reply]
Glad to hear it worked for you!  I know they've reduced the number of hooks available in v5.  The contributors have said why in other posts, but I forget where at the moment.  

Until documentation is done I've taken to reading the source code and found the hooks property on the server contract.

**Comment by skrenek** [comment]
I don't know if it's best practice or not, but you could implement a server-level middleware that performs logging "downstream" of the request (after the next()).

**Reply by adufr** [reply]
I've tried, but I didn't manage to catch errors this way :disappointed:
